src/lib.rs

```diff
@@ -1728,6 +1729,49 @@ pub mod __rt {
             }
         }
     }
+
+    /// An internal helper struct for usage in `#[wasm_bindgen(main)]`
+    /// functions to throw the error (if it is `Err`).
+    pub struct MainWrapper<T>(pub Option<T>);
+
+    pub trait Main {
+        fn __wasm_bindgen_main(&mut self);
+    }
+
+    impl Main for &mut &mut MainWrapper<()> {
+        #[inline]
+        fn __wasm_bindgen_main(&mut self) {}
+    }
+
+    impl Main for &mut &mut MainWrapper<Infallible> {
+        #[inline]
+        fn __wasm_bindgen_main(&mut self) {}
+    }
+
+    impl<E: Into<JsValue>> Main for &mut &mut MainWrapper<Result<(), E>> {
+        #[inline]
+        fn __wasm_bindgen_main(&mut self) {
+            if let Err(e) = self.0.take().unwrap() {
+                crate::throw_val(e.into());
+            }
+        }
+    }
+
+    impl<E: std::fmt::Debug> Main for &mut MainWrapper<Result<(), E>> {
+        #[inline]
+        fn __wasm_bindgen_main(&mut self) {
+            if let Err(e) = self.0.take().unwrap() {
+                crate::throw_val(std::format!("{:?}", e).into());
+            }
+        }
+    }
+
+    impl<T: std::process::Termination> Main for MainWrapper<T> {
+        #[inline]
+        fn __wasm_bindgen_main(&mut self) {
+            crate::throw_val(std::format!("{:?}", self.0.take().unwrap().report()).into());
```

Comment:
I didn't really mind this because exit codes don't really make sense in this context, so having every exit code fail sounded appropriate to me.

In retrospect that wasn't really a good idea at all, so I would argue to drop support for `Termination` all-together.

WDYT?

Reply:
> In retrospect that wasn't really a good idea at all, so I would argue to drop support for `Termination` all-together.

Okay, I'm fine with that.

> I thought about how we could address this in the future. The only way we could fix this for async functions, I can think of, is to introduce a `block_on` functionality, like in [`futures`](https://docs.rs/futures/0.3.28/futures/executor/fn.block_on.html) and [`tokio`](https://docs.rs/tokio/1.27.0/tokio/runtime/struct.Runtime.html#method.block_on), which might be possible with [`Atomic.waitAsync`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/waitAsync). Unfortunately this isn't available in Firefox yet ([bugzilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1467846))

Standard implementations of `block_on` using `std::thread::park` already work fine on wasm, but the issue with that is that the main thread on the web can't block (trying to use `Atomics.wait` or wasm's various atomic waiting instructions will fail), plus you need to yield back to the browser event loop for any `Promise`s to get resolved.

Reply:
> Standard implementations of `block_on` using `std::thread::park` already work fine on wasm, but the issue with that is that the main thread on the web can't block

Yes, that's the idea behind using `Atomic.waitAsync`, because `std::thread::park` would fail on the main thread.
> plus you need to yield back to the browser event loop for any `Promise`s to get resolved.

I'm not 100% sure on this, but `Atomic.waitAsync` should not prevent `Promise`s to get resolved, because `spawn_local` or `Promise`s in general are a different context and should not be blocked by `Atomic.waitAsync`.

Reply:
But the whole point of `Atomics.waitAsync` is that it doesn't block; it returns a `Promise`.

Reply:
Ah, right, I remembered wrongly, for some reason in my mind it just blocked in an async way, basically continuing the function when the `Promise` resolves.

Reply:
Alright, I removed `Termination` support, I tried my best re-formulating the documentation, feel free to suggest any better wording of it.